lms/static/completion/js/ViewedEvent.js

```diff
@@ -0,0 +1,182 @@
+    const now = Date.now();
+    if (time + wait < now) {
+      time = now;
+      fn();
+    }
+  }
+  return delay;
+}
+
+
+export class ElementViewing {
+  /**
+   * A wrapper for an HTMLElement that tracks whether the element has been
+   * viewed or not.
+   */
+  constructor(el, viewedAfterMs, callback) {
+    this.el = el;
+    this.viewedAfterMs = viewedAfterMs;
+    this.callback = callback;
+
+    this.topSeen = false;
+    this.bottomSeen = false;
+    this.seenForMs = 0;
+    this.becameVisibleAt = undefined;
+    this.hasBeenViewed = false;
+  }
+
+  getBoundingRect() {
+    return this.el.getBoundingClientRect();
+  }
+
+  /** This element has become visible on screen.
+   *
+   * (may be called even when already on screen though)
+   */
+  handleVisible(now) {
+    if (!this.becameVisibleAt) {
+      this.becameVisibleAt = now;
+      // We're now visible; after viewedAfterMs, if the top and bottom have been
+      // seen, this block will count as viewed.
+      setTimeout(
+        () => {
+          this.checkIfViewed(Date.now());
+        },
+        this.viewedAfterMs - this.seenForMs,
+      );
+    }
+  }
+
+  handleNotVisible(now) {
+    if (this.becameVisibleAt) {
+      this.seenForMs = now - this.becameVisibleAt;
+    }
+    this.becameVisibleAt = undefined;
+  }
+
+  markTopSeen(now) {
+    // If this element has been seen for enough time, but the top wasn't visible, it may now be
+    // considered viewed.
+    this.topSeen = true;
+    this.checkIfViewed(now);
+  }
+
+  markBottomSeen(now) {
+    this.bottomSeen = true;
+    this.checkIfViewed(now);
+  }
+
+  getTotalTimeSeen(now) {
+    if (this.becameVisibleAt) {
+      return this.seenForMs + (now - this.becameVisibleAt);
+    }
+    return this.seenForMs;
+  }
+
+  areViewedCriteriaMet(now) {
+    return this.topSeen && this.bottomSeen && (this.getTotalTimeSeen(now) >= this.viewedAfterMs);
+  }
+
+  checkIfViewed(now) {
+    // User can provide a "now" value for testing purposes.
+    if (this.hasBeenViewed) {
+      return;
+    }
+    if (this.areViewedCriteriaMet(now)) {
+      this.hasBeenViewed = true;
+      // Report to the tracker that we have been viewed
+      this.callback(this.el, { elementHasBeenViewed: this.hasBeenViewed });
+    }
+  }
+}
+
+
+export class ViewedEventTracker {
+  /**
+   * When the top or bottom of an element is
+   * first viewed, or the entire element is viewed for a specified amount
+   * of time, the callback is called, passing the element that was viewed,
+   * and an event object having the following field:
+   *
+   * *   hasBeenViewed (bool): true if all the conditions for being
+   *     considered "viewed" have been met.
+   */
+  constructor(elements, viewedAfterMs) {
+    this.viewedAfterMs = viewedAfterMs;
+    this.elementViewings = new Set();
+    this.visible = new Set();
+    this.previouslyVisible = new Set();
+    this.handlers = [];
+
+    this.interval = undefined;
+    this.registerDomHandlers();
```

Comment:
This needs to be moved to happen after the `ElementViewing` objects get created below... otherwise the initial call to `updateVisible()` in `registerDomHandlers()` won't do anything.

Or, move the initial call to `updateVisible()` to the end of this method, etc.